### DISCUSSION TITLE
SearXNG works from terminal but not as an uWSGI daemon

### DISCUSSION THREAD

**Discussion by gkkulik** [searxng/searxng#3052] [Q&A]
Hello, I installed SearXNG after wiping out my old installation. I used the installation script. Now I can run my instance from Terminal as [described in the documentation](https://docs.searxng.org/admin/installation-searxng.html#check) but it won't work as an uWSGI daemon. uWSGI doesn't output any errors, the entire log looks as if SearXNG was running correctly. Am I missing something?

Here's the `inspect` output:

```
uWSGI:
  SEARXNG_UWSGI_SOCKET : /usr/local/searxng/run/socket
environment /usr/local/searxng/searxng-src/utils/brand.env:
  GIT_URL              : https://github.com/searxng/searxng
  GIT_BRANCH           : master
  SEARXNG_URL          : http://user/searxng
  SEARXNG_PORT         : 8888
  SEARXNG_BIND_ADDRESS : 127.0.0.1
INFO:  uWSGI socket is located at: /usr/local/searxng/run/socket
INFO:  got 301 from http://user/searxng
```

Here's the `check` output:

```
SearXNG checks
--------------
INFO:  [OK] (old) account 'searx' does not exists
INFO:  [OK] (old) account 'filtron' does not exists
INFO:  [OK] (old) account 'morty' does not exists
INFO    searx                         : merge the default settings ( /usr/local/searxng/searxng-src/searx/settings.yml ) and the user settings ( /etc/searxng/settings.yml )
INFO    searx                         : max_request_timeout=None
INFO    searx.redisdb                 : connecting to Redis db=0 path='/usr/local/searxng-redis/run/redis.sock'
INFO    searx.redisdb                 : connected to Redis
```

Here's the `curl --location --verbose --head --insecure 127.0.0.1:8888` output:

```
*   Trying 127.0.0.1:8888...
* connect to 127.0.0.1 port 8888 failed: Connection refused
* Failed to connect to 127.0.0.1 port 8888 after 0 ms: Connection refused
* Closing connection 0
curl: (7) Failed to connect to 127.0.0.1 port 8888 after 0 ms: Connection refused
```

**Comment by virtadpt** [comment]
What does your uwsgi.ini file look like?

**Reply by gkkulik** [reply]
You mean the one in dockerfiles/? I don't use docker, just a regular installation. The documentation mentions /etc/uwsgi.ini but there is no file there. Is there any other place I should look into?

**Reply by virtadpt** [reply]
That there isn't a uwsgi.ini file suggests the problem: There needs to be one for uwsgi to know what to do.

Here's what mine looks like:

```
[uwsgi]
workers = 12
single-interpreter = true
master = true
plugin = python
lazy-apps = true
enable-threads = true
module = searx.webapp
virtualenv = /home/virtadpt/searxng/env/
pythonpath = /home/virtadpt/searxng/
chdir = /home/virtadpt/searxng/searx/
http = 127.0.0.1:8888
disable-logging = true
route-run = fixpathinfo:
cache2 = name=searxcache,items=2000,blocks=2000,blocksize=4096,bitmap=1
```

In a shell script (because I use a venv with SearxNG and systemd) I have the command `uwsgi --ini searxng.ini to start it up.

It sounds like you hit a deficiency in the docs.  I set this up so long ago I don't remember the specifics, but I do recall building the .ini file out little by little until I had something that works.

**Comment by gkkulik** [comment]
Okay, I tried to install SearXNG on a completely new system (Ububntu 22.04.3) and the problem is exactly the same. I've been trying to troubleshoot this for two days and I'm giving up. I know this is an open source effort and I'm one of the translators myself but this isn't worth it. Sorry, guys. Good luck.

**Reply by return42** [reply]
> Hello, I installed SearXNG after wiping out my old installation. I used the installation script. 

By default, the script installs uWSGI which listens on a socket ..

```
INFO:  uWSGI socket is located at: /usr/local/searxng/run/socket
```

> Here's the `curl --location --verbose --head --insecure 127.0.0.1:8888` output:

Since it is listen on socket, this won't work .. 

```
INFO:  got 301 from http://user/searxng
```

This URL does not exists .. you need to setup a HTTP server to proxy to the socket.

https://docs.searxng.org/admin/installation-scripts.html

![image](https://github.com/searxng/searxng/assets/554536/5bbfc79a-b5c4-41b3-8ffb-42dd0396db10)

If you prefer NGINX use ..

![image](https://github.com/searxng/searxng/assets/554536/dd76df61-0f4f-4994-a771-267c2a2125aa)

the script from above will install the proxy:

![image](https://github.com/searxng/searxng/assets/554536/ba28c274-faae-4e04-b73a-74021b3e8b3c)

In your network, the SearXNG instance should be available under:

    http://<hostname>/searxng

**Reply by gkkulik** [reply]
Thanks but the documentation says:

> **When all services are installed and running fine**, you can add SearXNG to your HTTP server.

How do I check if they are running fine then? There is no information about it in the documentation. Furthermore, if I get 301, they are obviously not running fine, are they? That part of the documentation is just badly written then.

Just below that it says: 

> We use caddy in our **[docker image](https://docs.searxng.org/admin/installation-docker.html#installation-docker)** and we have implemented installation procedures for

Correct me if I'm wrong but if someone uses the installation script, it's obvious they are not using a docker image, is it? What's the point of that information then? To me it implies that the installation procedures are for docker images, not for me.

**Reply by return42** [reply]
> Correct me if I'm wrong but if someone uses the installation script, it's obvious they are not using a docker image, is it? What's the point of that information then? To me it implies that the installation procedures are for docker images, not for me.

Hm, now when you say it I realized it .. yes, sorry it is misleading .. fixes are welcome.

> How do I check if they are running fine then?

Your "SearXNG checks" does not report an issue .. so everything seems fine so far.

>  Furthermore, if I get 301, they are obviously not running fine, are they?

```
INFO:  uWSGI socket is located at: /usr/local/searxng/run/socket
INFO:  got 301 from http://user/searxng
```

Understand the 301 as a request that HTTP must be installed.

It might be a bit confusing at this point of the installation procedure that http is tested before installed .. everything can be improved .. fixes are welcome.
